drafts/swift/redux-in-swift-pt-1.md

```diff
@@ -0,0 +1,280 @@
+    - This is partially why we need that `!` when defining the `state` in the `Store`.
+- They should return the current *State* if they are given an *Action* they cannot handle.
+
+### Dispatching Actions and Notifying Subscribers
+
+An *Action* can be dispatched to a *Store* to change its *State*. After a *Store*'s *State* is changed, the *Store* should notify all of its *Subscribers*. Thus,
+
+- A *Subscriber* is a function takes a *Store*
+- *Subscribers* can subscribe to a *Store* (will be explained in the next section)
+- A *Store* need to hold onto an array of *Subscribers*
+- After a *Store*'s *State* changes, the *Store* need to notify the change to all of its *Subscribers*
+
+```swift
+class Store<State> {
+    // ... previous code ...
+    typealias Subscriber = (store: Store) -> ()
+    final var subscribers = [Subscriber]()
+
+    final func dispatch(action: ActionType) {
+        self.state = reducer(state: state, action: action)
+        subscribers.forEach {
+            $0(store: self)
+        }
+    }
+}
+```
+
+Now the generic class `Store` is almost usable except a few pieces missing. Example usage:
+
+```swift
+struct Increase: ActionType { }
+
+let counterStore = Store<Int>.init { (state: Int?, action: ActionType) -> Int in /* ... */ }
+
+counterStore.dispatch(Increase())
+```
+
+If you want to run above snippet with minimum change, replace the `/* ... */` with `return 0`.
+
+#### How does a Store Get its Initial State?
+
+We need to update the `init` method in `Store` to the following:
+
+```swift
+class Store<State> {
+    // ... previous code ...
+    init(with reducer: Reducer) {
+        self.reducer = reducer
+
+        dispatch(InitialAction())
+    }
+}
+```
+
+and also add the definition of the `InitialAction` (before and outside the definition of the `Store`),
+
+```swift
+struct InitialAction: ActionType { }
+```
+
+So when a *Store* initializes, the following would happen regarding its `state`:
+
+- the `Store` will call its `dispatch` with the `InitialAction`
+- `dispatch` would update the `Store`'s `state` by calling its `reducer` with `nil` as the `state` param and `InitialAction` 
+- the `state` in the `Store` would be set to whatever the `Reducer` returns as the initial state
+    - because a `Reducer` returns the actual initial state when it receives a `nil` as its `state` param
+
+#### The BANG`!`
+
+As you can see from the above and actually running the snippet, the forced unwrapped `state` in the `Store` will always be set to non nil values before it's used as a non nil value. It is not only safe but also convenient to use (no need for `guard let`s everywhere).
+
+### Subscribing to a Store
+
+*Subscriber*s are functions which takes a *Store* as its only param and they can subscribe to *Store*s by calling the following method. When it's subscribed to a *Store*, it should get the *Store*'s current *State*.
```

Comment:
*...by calling the `subscribe` method as illustrated below. When a *Subscriber* is subscribed to a *Store*, it will receive the *Store*'s current *State*.